<a href="https://colab.research.google.com/github/Anastasia-Lyashuk/Portfolio/blob/main/bank_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1. Откроем таблицу и изучим общую информацию о данных

### 1.1. Импорт библиотеки pandas. Считайем данные из csv-файла в датафрейм и сохраните в переменную `data`.

In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

### 1.2. Выведем первые 20 строчек датафрейма


In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

### 1.3. Выведем основную информацию о датафрейме


In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## 2. Предобработка данных

### 2.1. Удалим пропуски

In [ ]:
data.isna().sum() # выведем количество пропущенных значений для каждого столбца

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median() # заполниv пропуски медианным значением по каждому типу дохода

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median() # заполним пропуски в столбце медианными значениями по каждому типу занятости

In [ ]:
data.isna().sum() # проверим пропуски после обработки

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### 2.2. Обработаем аномальные значения

In [ ]:
data['days_employed'] = data['days_employed'].abs() # заменим все отрицательные значения положительными

In [ ]:
data.groupby('income_type')['days_employed'].agg('median') # выведите медианное значение трудового стажа в днях.

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно и этот столбец не понадобится для дальнейшего исследования, поэтому оставим как есть

In [ ]:
data['children'].unique() # выведем уникальные значеня столбца

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)] # удалим строки, в которых встречаются аномальные значения из датафрейма

In [ ]:
data['children'].unique() # выведем уникальные значеня столбца для проверки

array([1, 0, 3, 2, 4, 5])

### 2.3. Изменим типы данных

In [ ]:
data['total_income'] = data['total_income'].astype(int) #  заменим вещественный на целочисленный тип данных

### 2.4. Обработаем дубликаты

In [ ]:
data['education'] = data['education'].str.lower() # приведем значения в столбце к нижнему регистру

In [ ]:
data.duplicated().sum() # выведем на экран количетво строк-дубликатов

71

In [ ]:
data = data.drop_duplicates() # удалим строки-дубликаты

### 2.5. Категоризация данных

#### 2.5.1. Создадим столбец total_income_category, в котором укажем категории, в зависимости от дохода.



In [ ]:
# создадим функцию для создания категорий в зависимости от размера дохода
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [ ]:
data['purpose'].unique() # выведем на экран перечень уникальных целей взятия кредита

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

#### 2.5.2. Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:
- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.


In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### 3. Исследовательский анализ данных


#### 3.1 Определим, есть ли зависимость между количеством детей и возвратом кредита в срок

In [ ]:
def children_quality(children):
    if children == 0:
        return 'бездетные заемщики'
    if children >= 3:
        return 'многодетные заемщики'

    return 'малодетные заемщики'

data['child'] = data['children'].apply(children_quality)
data_children_pivot = data.pivot_table(index=['child'],values='debt',aggfunc=['count','sum','mean'])
print(data_children_pivot)

                      count   sum      mean
                       debt  debt      debt
child                                      
бездетные заемщики    14091  1063  0.075438
малодетные заемщики    6860   638  0.093003
многодетные заемщики    380    31  0.081579


Самая высокая доля невозврата кредита у малодетных заемщиков (1-2 ребенка) и составляет 9.3%
Самая низкая доля невозврата кредита у бездетных заемщиков


#### 3.2 Определим, есть ли зависимость между семейным положением и возвратом кредита в срок.

In [ ]:
data_family_status_pivot = data.pivot_table(index=['family_status'], values='debt',aggfunc=['count','sum','mean'])
print(data_family_status_pivot)
 # Ваш код будет здесь. Вы можете создавать новые ячейки.

                       count  sum      mean
                        debt debt      debt
family_status                              
Не женат / не замужем   2796  273  0.097639
в разводе               1189   84  0.070648
вдовец / вдова           951   63  0.066246
гражданский брак        4134  385  0.093130
женат / замужем        12261  927  0.075606


Самая низкая доля невозврата в категориях "вдова/вдовец"
и "в разводе"  6.62% и 7.06% соответственно.
Самая высокая доля в категориях "не женат/не замужем" и "гражданский брак"  9.76% и 9.31% сответсвенно.  

#### 3.3 Определим, есть ли зависимость между уровнем дохода и возвратом кредита в срок

In [ ]:
data.groupby('total_income_category')['debt'].agg(['count','sum','mean'])
# Ваш код будет здесь. Вы можете создавать новые ячейки.

count   sum      mean
total_income_category                       
A                         25     2  0.080000
B                       5014   354  0.070602
C                      15921  1353  0.084982
D                        349    21  0.060172
E                         22     2  0.090909

In [ ]:
def categorize_income_new(income):
    try:
        if 0 <= income <= 80000:
            return 'доход менее 100000'
        elif 80001 <= income <= 200000:
            return 'доход от 100001 до 200000'
        elif income >= 200001:
            return 'доход более 200001'
    except:
        pass

data['total_income_category_new'] = data['total_income'].apply(categorize_income_new)
data.groupby('total_income_category_new')['debt'].agg(['count','sum','mean'])

count   sum      mean
total_income_category_new                       
доход более 200001          5039   356  0.070649
доход менее 100000          2265   174  0.076821
доход от 100001 до 200000  14027  1202  0.085692

Для анализа используются только категории В и С, поскольку они являются самыми крупными, а по остальным категориям недостаточно данных для корректных выводов. Таким образорм, самая высокая доля невозврата в категории С (доход от 50000 до 200000) 8.5%. Самая низкая доля невозврата в категории В (доход от 200000 до 1000000) 7.06%

#### 3.4 Выясним, как разные цели кредита влияют на его возврат в срок

In [ ]:
data.groupby('purpose_category')['debt'].agg(['count','sum','mean'])
# Ваш код будет здесь. Вы можете создавать новые ячейки.

count  sum      mean
purpose_category                              
операции с автомобилем     4279  400  0.093480
операции с недвижимостью  10751  780  0.072551
получение образования      3988  369  0.092528
проведение свадьбы         2313  183  0.079118

Самая высокая доля невозврата кредита в категорях "операции с автомобилем" 9.34% и "получение образования" 9.25%. В обеих категориях риск примерно одинаковый.
Самая низкая доля невозврата кредита в категориях "проведение свадьбы" 7.91% и "оперции с недвижимостью" 7.26%. В обоих категориях риск примерно одинаковый.

### 4. Общий вывод.

По итогам исследования были изучены некоторые факторы, влияющие на невозврат крелита, такие как количество детей, семейное положение, уровень дохода и цель кредита.

Самая высока доля невозврата кредита была выявлена в категорях "малодетные заемщики" 9.3%, "не женаты/не замужние" 9,76% и "в гражданском браке" 9.31%, а так же по опрециям с авто 9.35% и по образованию 9.25%. Доля невозврата кредита является самой высокой в категории С "доход от 50000 до 200000" и составляет 8,50%.
    

    
В ходе исследования была осуществлена категоризация данны, их сортировака и анализ по заданным критериям.
    
Рекомендации:
При выдаче кредита особое внимание обратить на котегории граждан, у которых 1 или два ребенка, не женатых/ не замужних или находящихся в гражданском баке, с доходом от 50000 до 200000, берущие кредит на учебу или на авто. При общении с потенциальным клиентом, который относится к двум и более перечисленным категориям, следут более внимательно проверить его платежеспособнось и возможно принять меры для уменьшения рисков для банка (возможно, не использовать льготную ставку и проч.)
